interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,877 @@
+		Constraints: constraints,
+		Outcome:     outcome,
+		Lifespan:    lifespan,
+		Expiration:  expiration,
+	}
+	return &newRule, nil
+}
+
+// IsPathAllowed checks whether the given path with the given permission is
+// allowed or denied by existing rules for the given user, snap, and interface.
+// If no rule applies, returns ErrNoMatchingRule.
+func (rdb *RuleDB) IsPathAllowed(user uint32, snap string, iface string, path string, permission string) (bool, error) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	needToSave := false
+	defer func() {
+		if needToSave {
+			rdb.save()
+		}
+	}()
+	variantMap := rdb.permissionDBForUserSnapInterfacePermission(user, snap, iface, permission).VariantEntries
+	matchingVariants := make([]patterns.PatternVariant, 0)
+	// Make sure all rules use the same expiration timestamp, so a rule with
+	// an earlier expiration cannot outlive another rule with a later one.
+	currTime := time.Now()
+	for variantStr, variantEntry := range variantMap {
+		matchingRule, err := rdb.ruleWithID(variantEntry.RuleID)
+		if err != nil {
+			// Database was left inconsistent, should not occur
+			delete(variantMap, variantStr)
+			// Record a notice for the offending rule, just in case
+			rdb.notifyRule(user, variantEntry.RuleID, nil)
+			continue
+		}
+		expired, err := matchingRule.Expired(currTime)
+		switch {
+		case err != nil:
+			// Should not occur
+			logger.Noticef("error while checking whether rule had expired: %v", err)
+			fallthrough
+		case expired:
+			needToSave = true
+			rdb.removeRuleFromTree(matchingRule)
+			rdb.removeRuleWithID(variantEntry.RuleID)
+			data := map[string]string{"removed": "expired"}
+			rdb.notifyRule(user, variantEntry.RuleID, data)
+			continue
+		}
+		// Need to compare the path pattern variant, not the rule's path
+		// pattern, so that only variants which match are included,
+		// and the highest precedence variant can be computed.
+		matched, err := patterns.PathPatternMatches(variantStr, path)
+		if err != nil {
+			// Only possible error is ErrBadPattern, which should not occur
+			return false, err
+		}
+		if matched {
+			matchingVariants = append(matchingVariants, variantEntry.Variant)
+		}
+	}
+	if len(matchingVariants) == 0 {
+		return false, ErrNoMatchingRule
+	}
+	highestPrecedenceVariant, err := patterns.HighestPrecedencePattern(matchingVariants, path)
+	if err != nil {
+		return false, err
+	}
+	matchingEntry := variantMap[highestPrecedenceVariant.String()]
+	matchingID := matchingEntry.RuleID
+	matchingRule, err := rdb.ruleWithID(matchingID)
+	if err != nil {
+		// Database was left inconsistent, should not occur
+		return false, ErrRuleIDNotFound
+	}
+	if matchingRule.Lifespan == prompting.LifespanSingle {
+		// XXX: we should never add rules with lifespan single to the rule DB
+		rdb.removeRuleFromTree(matchingRule)
+		rdb.removeRuleWithID(matchingID)
+		data := map[string]string{"removed": "expired"}
+		rdb.notifyRule(user, matchingID, data)
+		needToSave = true
```

Comment:
can you explain what this code is about?

Reply:
This code is about removing any `"lifespan": "single"` rule from the database the first time it is used to match a request. Since we should never create a rule for a reply with `"lifespan": "single"`, we shouldn't have any in the rule DB, unless the client explicitly adds or patches a rule to have lifespan "single". But we should probably not let the client do this -- note to self: add this check and send back `BadRequest` if the client does so.

Reply:
As part of disallowing the creation of rules with lifespan `"single"`, this code was removed.